spmd/tensor/device_mesh.py

```diff
@@ -531,102 +460,87 @@ def reduce_scatter(
         Returns:
             A :class:`torch.Tensor` object
         """
-        my_coordinate = self.get_coordinate_on_dim(mesh_dim)
-        # TODO: what should happen if rank is not in the mesh?
-        # see issue https://github.com/pytorch/tau/pull/492
-        assert (
-            my_coordinate is not None
-        ), "Rank if not part of mesh"  # TODO: figure out behavior here
-
-        num_chunks = self.size(mesh_dim)
         if self._backend == "nccl":
-            input_list = list(input.tensor_split(num_chunks, dim=tensor_dim))
             # gloo does not support uneven scattering, nccl supports it, but it
             # might be slow compared to even size collective, we need to pad tensor
             # before really calling scatter, and unpad/narrow it after the collective
             # TODO: consider if we should remove this logic once ProcessGroupGloo
             # suupport uneven list, and collective perfomance on par
-            idx_start_to_pad = input.size(tensor_dim) % num_chunks
-            to_scatter = []
-            for i, scatter_tensor in enumerate(input_list):
-                if idx_start_to_pad != 0 and i >= idx_start_to_pad:
-                    scatter_tensor = self._pad_tensor_dim_by_1(
-                        scatter_tensor, tensor_dim
-                    )
-                scatter_tensor = scatter_tensor.contiguous()
-                to_scatter.append(CommTensor(scatter_tensor))
-
-            output = torch.empty_like(to_scatter[my_coordinate])
             dim_group = self._dim_groups[mesh_dim]
-            reduce_scatter(output, to_scatter, op=op, group=dim_group)
+            fut = reduce_scatter(output, input_list, op=op, group=dim_group)
 
-            # resize to uneven size if needed
-            if idx_start_to_pad != 0 and my_coordinate >= idx_start_to_pad:
-                # tensor = tensor.narrow(tensor_dim, start=0, length=tensor.size(tensor_dim) - 1)
-                output = self._unpad_tensor_dim_by_1(output, tensor_dim)
-
-            return output
         elif self._backend == "gloo":
             # it's gloo, which does not have reduce_scatter
             # we have to do all_reduce + scatter
```

Comment:
This is probably premature optimization but IIUC:

For `mesh_dim` equal to `k` and assuming equal-sized size-`n`-byte tensors in `tensor_list` , the current gloo mesh reduce-scatter allocates `(k-1) * n` bytes (via `clone()`); copies `k * n` bytes (`n` via `copy_()` and `(k-1) * n` via `clone()`); and calls `k` all-reduces.

I think by increasing the memory usage slightly, you may decrease the number of all-reduces to just `1`: You allocate `k * n` bytes at once in a single tensor (i.e. `torch.empty()`), copy `k * n` bytes into tensor (i.e. the `input_list` contents), call `1` all-reduce, and copy `n` bytes into `output`. This allocates the same number of bytes and requires copying `n` additional bytes but makes the number of all-reduces a constant `1`.

I am not sure how the single `n * k`-byte tensor allocation compares to `k` separate `n`-byte allocations in terms of memory fragmentation,. but I hypothesize that for small and medium sizes (which DT may target), the `n * k` byte allocation will be better.

Reply:
Thanks for the suggestion! Yeah this looks better, will update.